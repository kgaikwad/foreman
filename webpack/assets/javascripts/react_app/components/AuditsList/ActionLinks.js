import $ from 'jquery';
import React from 'react';
import { Col } from 'patternfly-react';
import { OverlayTrigger, Tooltip } from 'react-bootstrap';
import { notify } from '../../../foreman_toast_notifications';

const notifyOnError = (responseText) => {
  try {
    const response = JSON.parse(responseText);
    notify({ message: `<p>${response.message}</p>`, type: 'error' });
  } catch (ex) {
    notify({ message: `<p>${responseText}</p>`, type: 'error' });
  }
};

class ActionLinks extends React.Component {
  constructData(dataInfo) {
    const changedData = {};
    const auditedChanges = Object.entries(this.props.auditedChanges);
    if (auditedChanges.length) {
      auditedChanges.forEach(([key, value]) => {
        [changedData[key]] = value;
      });
    }
    if (dataInfo.id) {
      changedData.id = dataInfo.id;
    }
    return changedData;
  }

  handleClick(dataInfo, e) {
    if (!$.rails.confirm(__('Do you really want to revert these changes?'))) {
      return;
    }
    e.preventDefault();

    if (dataInfo.url) {
      $.ajax({
        url: dataInfo.url,
        data: JSON.stringify(this.constructData(dataInfo)),
        type: 'PUT',
        contentType: 'application/json',
        dataType: 'json',
        processData: false,
        success(response, textStatus, xhr) {
          notify({ message: `<p>${response.message}</p>`, type: 'success' });
          window.Turbolinks.visit(response.success_redirect);
        },
        error(xhr, textStatus, error) {
          notifyOnError(xhr.responseText);
        },
      });
    }
  }

  render() {
    return (
      <Col sm={2} className='actions-btns'>
        { this.props.allowedActions && this.props.allowedActions.map((action, index) => {
          const tooltip = <Tooltip id={index}>{action.title}</Tooltip>;

          const linkProp = {
            className: action.css_class,
            href: action.url,
            disabled: action.disabled,
          };

          return (
            <OverlayTrigger placement="top" id="notifications-toggle-icon"
              overlay={tooltip} key={index}>
              <a key={index} {...linkProp}
                 onClick={action.data ? this.handleClick.bind(this, action.data) : null}>
                {action.name ? action.name : action.title}
              </a>
            </OverlayTrigger>
          );
        })}
      </Col>
    );
  }
}

export default ActionLinks;
