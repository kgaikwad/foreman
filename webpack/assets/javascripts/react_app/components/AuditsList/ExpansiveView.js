import React from 'react';
import { renderTemplatesDiff } from '../../../foreman_editor';

const showTemplateDiffIfAny = (auditTitle, template) => {
  if (template && template[0] !== template[1]) {
    const textareaProp = {
      label: __('Template diff'),
      'data-file-name': auditTitle,
      className: 'col-md-12 editor_source diffMode',
      type: 'text',
    };

    return (
      <div>
        <div className="editor-container">
          <textarea {...textareaProp} />
        </div>
        <input type="hidden" id="old" value={template[0]} />
        <input type="hidden" id="new" value={template[1]} />
      </div>
    );
  }
  return null;
};

const renderListItems = items =>
  items &&
  items.map((item, index) => (
    (item && typeof item === 'string' && item.length > 0) ? <li key={index}>{item}</li> : null
  ));

const renderCols = changeArr =>
  changeArr &&
  changeArr.map((change, index) => (
    <td key={index}><div className={change.css_class}><p>{change.id_to_label}</p></div></td>
  ));

const renderTableRows = changeEntries => (
  changeEntries &&
  changeEntries.map((propDetail, index) => (
    <tr key={index}>
      <td key={index}><div><p>{ propDetail.name }</p></div></td>
      { renderCols(propDetail.change) }
    </tr>
  )));

const showChangesDependingUponActions = (audit) => {
  if (['added', 'removed'].includes(audit.action_display_name)) {
    return (
      <div className={ `details-row change-list ${audit.action_display_name === 'added' ? 'show-new' : 'show-old'}` }>
        <ul>
          { renderListItems(audit.details) }
        </ul>
      </div>
    );
  }

  return (
    <table className="table-changes table table-bordered table-hover">
      <tbody>
        { renderTableRows(audit.audited_changes_with_id_to_label) }
      </tbody>
    </table>
  );
};

class ExpansiveView extends React.Component {
  componentDidMount() {
    renderTemplatesDiff(this.nv);
  }

  render() {
    const auditObj = this.props.audit;

    return (
      <div ref={(elem) => { this.nv = elem; }} className="grid-container">
        { showTemplateDiffIfAny(auditObj.audit_title, auditObj.audited_changes.template) }
        { showChangesDependingUponActions(auditObj) }
        {
          auditObj.comment &&
            <div className="details-row comment-section">
              <p className='comment-title'><strong>{ __('Comments') }</strong></p>
              <p className='comment-desc'>{ auditObj.comment }</p>
            </div>
        }
      </div>
    );
  }
}

export default ExpansiveView;
