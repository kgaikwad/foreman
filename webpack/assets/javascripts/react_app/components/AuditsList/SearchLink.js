import React from 'react';

class SearchLink extends React.Component {
  render() {
    const linkProps = {
      href: this.props.url,
      title: this.props.title,
      'data-toggle': 'tooltip',
      'data-placement': 'bottom',
      id: `#resource-link-${this.props.id}`,
    };

    return (
      <a { ...linkProps }>{this.props.textValue}</a>
    );
  }
}

export default SearchLink;
