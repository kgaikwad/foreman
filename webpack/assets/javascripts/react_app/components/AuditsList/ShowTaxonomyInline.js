import React from 'react';
import { Row, Col } from 'patternfly-react';

const ShowTaxonomyInline = ({
  displayLabel = '',
  items = [],
}) => {
  const listItems = items.map((item, index) => (
    <a href={item.url} key={index} className="apply-comma {item.css_class}" disabled={item.disabled}>{item.name}</a>
  ));

  return (
    <Row>
      <span className="col-md-2">{displayLabel}</span>
      <Col sm={10}>
        <strong>
          {items && listItems}
        </strong>
      </Col>
    </Row>
  );
};

export default ShowTaxonomyInline;
