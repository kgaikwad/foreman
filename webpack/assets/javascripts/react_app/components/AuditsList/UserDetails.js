import React from 'react';

class UserDetails extends React.Component {
  render() {
    const {
      isAuditLogin,
      userInfo,
      remoteAddress,
    } = this.props;

    const {
      search_path: searchPath,
      display_name: UserDisplayName,
      audit_path: auditPath,
    } = userInfo;

    const linkProps = {
      href: searchPath,
      title: __('Filter audits for this user only'),
      'data-toggle': 'tooltip',
      'data-placement': 'bottom',
      className: 'user-info',
    };

    if (isAuditLogin) {
      return (
        <div>
          <a { ...linkProps }>{UserDisplayName}</a>
          <a href={auditPath }>__('Logged-in')</a>
        </div>
      );
    }

    return (
      <div>
        <a { ...linkProps }>{UserDisplayName}</a>
        {remoteAddress ? (<p className='gray-text'>({remoteAddress})</p>) : null}
      </div>
    );
  }
}

export default UserDetails;
