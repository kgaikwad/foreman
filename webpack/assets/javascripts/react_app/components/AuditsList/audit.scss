#audit-list {
  .apply-comma:not(:last-child)::after {
    white-space: pre;
    content: ", ";
  }

  .list-view-pf-body > .list-view-pf-description {
    width: 30%;
    flex: 1 0 30%;
  }

  .list-group-item-header {
    .list-group-item-heading {
      flex: 1 0 unquote("calc(45% - 20px)");
      width: unquote("calc(45% - 20px)"); // IE9 fallback
    }
  }

  .list-group-item-text {
    flex: 1 0 auto;
    width: calc(55% - 40px);
  }

  .list-group-item-container {
    padding-top: 20px;

    .actions-btns {
      padding-bottom: 30px;

      .btn {
        float: right;
        margin-left: 10px;
      }
    }

    .grid-container {
      .details-row {
        border: 1px solid #d1d1d1;
        margin-top: 15px;
      }

      .change-list {
        ul {
          margin-top: 10px;

          li {
            list-style-type: none;
          }
        }
      }

      .comment-section {
        p {
          margin-left: 10px;
          margin-right: 15px;
        }

        .comment-title {
          margin-top: 10px;
        }
      }

      .table-changes {
        margin-top: 15px;

        td {
          padding: 5px;
          background-color: #edf8ff;

          p {
            margin: 0 15px;
          }
        }
      }

      .table-hover > tbody > tr:hover > td,
      .table-hover > tbody > tr:hover > th {
        border: 1px solid #d1d1d1;
      }

      .show-old {
        background-color: #ffeef0;
      }

      .show-new {
        background-color: #e6ffed;
      }
    }
  }
}
