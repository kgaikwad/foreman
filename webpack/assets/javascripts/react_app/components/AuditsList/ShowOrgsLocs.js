import React from 'react';
import { Col } from 'patternfly-react';
import ShowTaxonomyInline from './ShowTaxonomyInline';

const ShowOrgsLocs = ({
  isOrgEnabled = false,
  isLocEnabled = false,
  orgs = [],
  locs = [],
}) => (
  <Col sm={10} className='taxonomy-section'>
    { isOrgEnabled &&
        <ShowTaxonomyInline displayLabel={__('Affected Organizations')} items={orgs}></ShowTaxonomyInline>
    }
    { isLocEnabled &&
        <ShowTaxonomyInline displayLabel={__('Affected Locations')} items={locs}></ShowTaxonomyInline>
    }
  </Col>
);

export default ShowOrgsLocs;
