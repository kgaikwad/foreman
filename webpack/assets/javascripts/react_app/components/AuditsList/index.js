import React from 'react';
import { ListView, Row } from 'patternfly-react';
import SearchLink from './SearchLink';
import ShowOrgsLocs from './ShowOrgsLocs';
import ActionLinks from './ActionLinks';
import ExpansiveView from './ExpansiveView';
import './audit.scss';

const isAuditLogin = (audit) => {
  let name;
  try {
    [name] = Object.keys(audit.audited_changes);
  } catch (e) {
    name = '';
  }
  return name === 'last_login_on';
};

const style = { color: '#999' };

const userDetails = (audit) => {
  const userInfo = audit.user_info;
  const linkProps = {
    href: userInfo.search_path,
    title: 'Filter audits for this user only',
    'data-toggle': 'tooltip',
    'data-placement': 'bottom',
    className: 'user-info',
  };
  if (isAuditLogin(audit)) {
    return (
      <div>
        <a { ...linkProps }>{userInfo.display_name}</a>
        <a href={userInfo.audit_path }>__('Logged-in')</a>
      </div>
    );
  }

  return (
    <div>
      <a { ...linkProps }>{userInfo.display_name}</a>
      {audit.remote_address ? (<p style={style}>({audit.remote_address})</p>) : null}
    </div>
  );
};

const description = audit => (
  <ListView.Description>
    <ListView.DescriptionText>{audit.action_display_name}</ListView.DescriptionText>
  </ListView.Description>
);

const renderAdditionalInfoItems = items =>
  items &&
  items.map((item, index) => (
    <ListView.InfoItem key={index}>
      {item}
    </ListView.InfoItem>
  ));

const renderActions = timeObject =>
  <span title={timeObject.title} style={style}>{timeObject.value}</span>;

const renderResourceLink = (audit) => {
  if (audit.audit_title_url) {
    return (
     <SearchLink url={audit.audit_title_url} textValue={audit.audit_title}
              title='Filter audits for this resource only' id={audit.id}></SearchLink>
    );
  }
  return audit.audit_title;
};

export default ({ data: { audits, isOrgEnabled, isLocEnabled } }) => (
  <ListView>
    {audits.map((audit, index) => (
      <ListView.Item id={audit.id} key={audit.id}
        actions={renderActions(audit.creation_time)}
        additionalInfo={
          renderAdditionalInfoItems([
            audit.audited_type_name.toUpperCase(), renderResourceLink(audit)])
        }
        heading={userDetails(audit)}
        description={description(audit)}
        stacked={false}
        hideCloseIcon={true}
      >
        <Row>
          <ShowOrgsLocs isOrgEnabled={isOrgEnabled} isLocEnabled={isLocEnabled}
            orgs={audit.affected_organizations} locs={audit.affected_locations} />
          <ActionLinks allowedActions={audit.allowed_actions}
                      auditedChanges={audit.audited_changes}/>
        </Row>

        <ExpansiveView audit={audit} />
      </ListView.Item>
    ))}
  </ListView>
);
