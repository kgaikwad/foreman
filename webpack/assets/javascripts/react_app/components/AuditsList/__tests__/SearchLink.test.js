import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import React from 'react';
import SearchLink from '../SearchLink';

import { searchLinkProp } from './AuditsList.fixtures';

describe('render search link', () => {
  it('should have render a search link', () => {
    const wrapper = shallow(<SearchLink { ...searchLinkProp }/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
