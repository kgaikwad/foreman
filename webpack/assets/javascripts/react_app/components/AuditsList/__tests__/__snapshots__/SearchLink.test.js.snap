// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render search link should have render a search link 1`] = `
<a
  data-placement="bottom"
  data-toggle="tooltip"
  href="/audits?search=type+%3D+user+and+auditable_id+%3D+1"
  id="#resource-link-123"
>
  testUser
</a>
`;
